nemo_aligner/models/nlp/gpt/megatron_gpt_spin_model.py

```diff
@@ -0,0 +1,632 @@
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import warnings
+from contextlib import nullcontext
+
+import torch
+from apex.transformer.pipeline_parallel.utils import get_num_microbatches
+from megatron.core import parallel_state
+from megatron.core.pipeline_parallel.schedules import get_forward_backward_func
+from omegaconf import OmegaConf
+from omegaconf.dictconfig import DictConfig
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.nlp.models.language_modeling.megatron_gpt_model import MegatronGPTModel
+from nemo.collections.nlp.modules.common.megatron.utils import (
+    average_losses_across_data_parallel_group,
+    get_iterator_k_split,
+)
+from nemo.collections.nlp.parts.utils_funcs import get_last_rank
+from nemo_aligner.models.alignable_interface import SupervisedInterface
+from nemo_aligner.utils.distributed import broadcast_2d_tensor, from_parallel_logits_to_logprobs
+from nemo_aligner.utils.train_utils import (
+    finish_validation_step,
+    grad_reductions,
+    prepare_for_training_step,
+    prepare_for_validation_step,
+    set_eval,
+    set_sync_funcs,
+    set_train,
+)
+from nemo_aligner.utils.utils import (
+    configure_batch_sizes,
+    cpu_weight_swap,
+    make_sharded_tensors_from_reference,
+    offload_distributed_adam,
+)
+
+
+class MegatronGPTSPINModel(MegatronGPTModel, SupervisedInterface):
+    """
+    Megatron GPT SPIN Model Training
+    Adapted from the paper Self-Play Fine-Tuning Converts Weak Language Models to Strong Language Models (Chen, et al, 2024)
+    https://arxiv.org/abs/2401.01335
+    
+    Our implementation differs in that we do not have a scheduler for the KL divergence parameter and we do not currently
+    inject generations from iteration t-1 into iteration t.
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer=trainer)
+        self.automatic_optimization = False
+
+        if self.cfg.pipeline_model_parallel_size > 1 and not self.cfg.megatron_amp_O2:
+            warnings.warn(
+                "when using pipeline parallelism, it is recommended to set megatron_amp_O2 to be True to "
+                "avoid explicit casting for pipeline communication"
+            )
+
+        self.ref_policy_state_dict = None
+        self.distributed_adam_offload_manager = None
+        self.to_offload_adam_states = self.cfg.spin.offload_adam_states
+
+        # self.ref_policy_kl_penalty = self.cfg.spin.get("ref_policy_kl_penalty", 0.0)
+        self.spin_config = OmegaConf.to_container(self.cfg.spin, resolve=True)
+        if isinstance(self.spin_config["ref_policy_kl_penalty"], float):
```

Comment:
Safer as some people may use `0` in the config:
```suggestion
        if isinstance(self.spin_config["ref_policy_kl_penalty"], (float, int)):
```

Reply:
Fixed, thanks